src/object.d

```diff
@@ -4750,16 +4749,22 @@ private void onArrayCastError()(string fromType, size_t fromSize, string toType,
     ];
 
     // convert discontiguous `msgComponents` to contiguous string on the stack
-    size_t length = 0;
-    foreach (m ; msgComponents)
-        length += m.length;
-
-    auto msg = (cast(char*)alloca(length))[0 .. length];
```

Comment:
Can't you just do 
```d
char[] msg;
if (__ctfe)
    msg = new char[length];
else
    msg = (cast(char*)alloca(length))[0 .. length];
```

or are you trying to get rid of the `alloca`?

Reply:
hmm.  I wonder if that would work.  Would `length` be determined at compile-time?

Reply:
I am trying to get rid of `alloca` in a way.  This was a response I received from @WalterBright  in a private e-mail exchange when I was trying to figure out how to create error messages dynamically in pure code.  

Walter's response to me.
> I understand the need to use alloca() to create error messages purely. But
there's another way:
>
> Use char[100]tmp; and simply truncate the message on overflow. It's not perfect,
one can't have a 5000 character error message, but it works for now, and gets
you unblocked.

I chose 2048 arbitrarily to ensure there was plenty of room for even the largest error message with templated types.

Reply:
Getting rid of `alloca` would also permit this to work in -betterC.  Currently `alloca` is compiled into druntime, so it's not linked in when compiling with -betterC.

Reply:
That seems fine.

>I chose 2048 arbitrarily to ensure there was plenty of room for even the largest error message with templated types.

In which case please reorder the message s.t. the array lengths are given first as they are bounded in length.

Reply:
Sorry, but I don't understand.  Reordering the strings would result in a grammatically incorrect sentence.

Reply:
Cannot cast array:  size 42, does not align on an array of size 21, of type foo, to type bar

Reply:
I went with "An array of size 42 does not align on an array of size 21, so `foo` cannot be cast to `bar`.  I hope that will work.

Reply:
Nice!